exp/runtime/api/v1beta1/extensionconfig_types.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// ANCHOR: ExtensionConfigSpec
+
+// ExtensionConfigSpec defines the desired state of ExtensionConfig.
+type ExtensionConfigSpec struct {
+	// ClientConfig defines how to communicate with ExtensionHandlers.
+	ClientConfig ClientConfig `json:"clientConfig"`
```

Comment:
This looks required, fields aren't required by default in open API so maybe we need `// +kubebuilder:validation:Required`?

Reply:
Q: Is this just to signal it in the code as well? If a field has neither +optional, pointer or omitempty it will be required in the CRD (aka it is required right now)

Reply:
You can also specify `// +kubebuilder:validation:Optional` on a package level, which then switches this defaulting round so that fields are optional if no tag is specified. If someone sets that on your package, it can be quite easy to miss, so it's better to be explicit about what you want on each field

Reply:
Sounds crazy to make all fields in a package optional, but I'm fine with making required explicit if others agree.

Reply:
> Sounds crazy to make all fields in a package optional

Which is why it's going to be very surprising if anyone ever did ;)

Reply:
I'd agree explicit is good but in this PR I'd suggest to be consistent with existing approach (which is not setting `Required` tag), then in a separate PR we can discuss and add that explicitly through the whole API if we want to.

Reply:
I've opened an issue to discuss this specifically here: https://github.com/kubernetes-sigs/cluster-api/issues/6397

I think if we decide this is a good idea we could definitely add the required tags to this API in a follow up. I'm not sure if adding it to existing APIs would count as a breaking change :thinking:.

Reply:
I agree with @enxebre, let's keep this consistent with the existing approach. 
From a quick test adding `+kubebuilder:validation:Required` doesn't affect the generated CRD (the field is already treated as required no matter of the marker), so with the current approach we are slightly less explicit in the code, but this won't affect kubectl explain.

Reply:
@JoelSpeed If you're happy to discuss the additional kubebuilder tags / OpenAPI validation on the parallel issues I'll resolve the conversations here.

Reply:
That's fine by me (note I'm now out of office until the 26th so responses will be slower)